processor/groupbytraceprocessor/processor.go

```diff
@@ -0,0 +1,328 @@
+
+	if !sp.ringBuffer.contains(traceID) {
+		// we likely received multiple batches with spans for the same trace
+		// and released this trace already
+		sp.logger.Debug("skipping the processing of expired trace", zap.Stringer("traceID", traceID))
+		return nil
+	}
+
+	// delete from the map and erase its memory entry
+	sp.ringBuffer.delete(traceID)
+
+	// this might block, but we don't need to wait
+	sp.logger.Debug("marking the trace as released", zap.Stringer("traceID", traceID))
+	go sp.markAsReleased(traceID)
+
+	return nil
+}
+
+func (sp *groupByTraceProcessor) markAsReleased(traceID pdata.TraceID) error {
+	// #get is a potentially blocking operation
+	trace, err := sp.st.get(traceID)
+	if err != nil {
+		return fmt.Errorf("couldn't retrieve trace %q from the storage: %w", traceID, err)
+	}
+
+	if trace.IsNil() {
+		return fmt.Errorf("the trace %q couldn't be found at the storage", traceID)
+	}
+
+	// signal that the trace is ready to be released
+	sp.logger.Debug("trace marked as released", zap.Stringer("traceID", traceID))
+
+	sp.cancelLock.RLock()
+	if !sp.cancelReleases {
+		sp.released <- trace
+		sp.removed <- traceID
+	}
+	sp.cancelLock.RUnlock()
+	return nil
+}
+
+func (sp *groupByTraceProcessor) onReleased() {
+	for item := range sp.released {
+		rs := item
+		trace := pdata.NewTraces()
+		trace.ResourceSpans().Append(&rs)
+		sp.nextConsumer.ConsumeTraces(context.Background(), trace)
+	}
+}
+
+func (sp *groupByTraceProcessor) onRemoved() {
+	for traceID := range sp.removed {
+		// this might block, but we don't need to wait
+		go sp.processRemoved(traceID)
+	}
+}
+
+func (sp *groupByTraceProcessor) processRemoved(traceID pdata.TraceID) error {
+	trace, err := sp.st.delete(traceID)
+	if err != nil {
+		return fmt.Errorf("couldn't delete trace %q from the storage: %w", traceID.String(), err)
+	}
+
+	if trace.IsNil() {
+		return fmt.Errorf("trace %q not found at the storage", traceID.String())
+	}
+
+	return nil
+}
+
+func (sp *groupByTraceProcessor) addSpans(traceID pdata.TraceID, trace pdata.ResourceSpans) error {
+	sp.logger.Debug("creating trace at the storage", zap.Stringer("traceID", traceID))
+	return sp.st.createOrAppend(traceID, trace)
+}
+
+func (sp *groupByTraceProcessor) processShutdown(ctx context.Context) {
+	sp.logger.Debug("closing processor")
+
+	// TODO: use the context to allow cancellation of the lock acquisition
+	// in theory, the lock should be released at the latest after sp.config.WaitDuration
+	sp.cancelLock.Lock()
+
+	sp.cancelReleases = true
+	close(sp.stopped)
+	close(sp.received)
+	close(sp.expired)
+	close(sp.removed)
+	close(sp.released)
+	sp.cancelLock.Unlock()
+}
+
+func extractTraceID(trace pdata.ResourceSpans) (pdata.TraceID, error) {
+	ilSpans := trace.InstrumentationLibrarySpans()
+	if ilSpans.Len() <= 0 {
+		return nil, errNoInstrumentationLibrarySpans
+	}
+
+	spans := ilSpans.At(0).Spans()
+	if spans.Len() <= 0 {
+		return nil, errNoSpans
+	}
+
+	return spans.At(0).TraceID(), nil
```

Comment:
Could spans in one slice have different traceIDs?

Reply:
Up to this moment, I was under the impression that `ResourceSpans` would hold only spans related to the same trace, but I just realized that my assumption might be wrong. 

https://github.com/open-telemetry/opentelemetry-proto/blob/dfa8c325321d55cd9e27fe6e2c7bf927bb4d9ca3/opentelemetry/proto/trace/v1/trace.proto#L27-L56

Reply:
I'm afraid you may need to group spans by traceID into separate batches and process each separately.

Reply:
@bogdandrutu could you please confirm which assumption is correct?